Add a method to query data using multiple paged queries

cvybhu opened the pull request:
## Description
This PR adds new methods: `query_all` and `execute_all`.  
They use multiple paged queries to fetch results, and then merge all those results into one.  
`page_size` must be set on the query that is passed to those methods.

Currently we use a single unpaged query to fetch data, but this might cause stalls (#298).
It's better to use many smaller queries and merge results locally.

```rust
Connection::query_all(
        &self,
        query: &Query,
        values: impl ValueList,
    ) -> Result<QueryResult, QueryError>;

Connection::execute_all(
        &self,
        prepared_statement: &PreparedStatement,
        values: impl ValueList,
    ) -> Result<QueryResult, QueryError>;
```

## Fixed Issues
Fixes: #298

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

Jasperav commented:
I thought the `execute_iter` gave back an iterator which lazy loads new page row results when the current page is exhausted, doesn't it do the same thing? It looks more memory efficient as well (if you have millions of rows, `execute_iter` won't care since it lazy loads the results and deallocates old rows). If you have millions of rows and you call this method, changes are you get out of memory exceptions, or am I misunderstanding this method?

Jasperav commented on scylla/src/transport/session.rs:
````diff
@@ -456,6 +457,71 @@ impl Session {
+    ///
+    /// # Examples
+    /// ```rust
+    /// # extern crate scylla;
````
The use of `extern crate` is from Rust 2015, it isn't needed anymore

cvybhu replied:
:+1:, removed.

psarna commented:
@Jasperav using iterators is absolutely preferred, especially for large datasets. This set of helper functions is mainly for internal use, where we know that the set is small, but we still want to use paging in order to avoid unnecessarily stressing the server - e.g. when fetching topology information.

cvybhu commented:
Ok @psarna @piodul I think it's ready for review

psarna commented:
@Jasperav ongoing. Note however, that this interface is intended to be used mostly for internal purposes, e.g. topology fetching. User data is often too large to gather everything in memory at one go (like these functions do), so it's **strongly** preferred to use our iterator interface. `query_all` and friends should only be used if you know that the result will be small and you don't want to worry about fetching multiple pages manually or advancing an iterator by hand.

piodul reviewed (changes requested):
I don't think that we should have `query_all` and `execute_all` in `Session`. The reason for that is that we already have `execute_iter` and `execute_all` which, as @Jasperav and @psarna mentioned, are more memory efficient and are preferred for users. I don't see a reason why we couldn't use it internally, too.

On the other hand, `Connection::query_all` and `Connection::execute_all` make more sense because, currently, the iterator interface needs the whole session to work (mainly for handling retries and routing, I think). When fetching the topology information we want to fetch pages using the control connection only - that's the main difference.

I'm fine with `query_all` and `execute_all` in `Connection`, but I wonder if it would be hard to write a `query_iter`/`execute_iter` which works on a single connection? Considering that the reason we wanted to add `query_all`/`execute_all` in the first place was preventing large allocations on the Scylla side, it would also be desirable to prevent large allocations on the client, too. If adapting the iterator is too hard, maybe  `query_all`/`execute_all` could receive a callback which would be called on each row?

Jasperav commented:
@piodul I am using this crate in my toy project and I actually needed `query_iter` on `Session`. It is a query on a small (<20 rows) set. If I don't have this method, I will end up writing my own implementation (probably initializing an empty vec, use the iterator to add rows). 

That is something this method already will do. I think other people sometimes just need all the rows in memory to send back to a client. So IMO this methods should remain on `Session`.

psarna commented on scylla/src/transport/errors.rs:
```diff
@@ -239,6 +239,14 @@ pub enum BadQuery {
+    #[error("Page size not set for the query when it should be.
+             Methods like query_all() differ from query() because they use paging to not overload the database.
+             Page size should be set using set_page_size()")]
+    PageSizeNotSet,
```
It's supposed to be used internally *for most cases*, but it does make some sense to also externalize the API, even for library coders (like @Jasperav) to have a neat API for reading stuff without having to manually collect an iterator each time.

We could consider adding a warning to this mechanism if the number of pages starts exceeding some small threshold (10? 100?), or when the result vector gets big enough (a couple of megabytes)

Jasperav commented:
The `try_collect` method bounds are Default and Extend, which are implemented for simple primitives in your example (i32, String) but not for every struct. Forcing clients to implement default and extend for types that will collect an async stream isn't desired IMO.

cvybhu commented:
v2:
* Removed Session::query_all, it's backed up on the [queryall-backup](https://github.com/cvybhu/scylla-rust-driver/tree/queryall-backup) branch in case we wanted it in the future

cvybhu commented:
v2:
* Removed Session::query_all, it's backed up on the [queryall-backup](https://github.com/cvybhu/scylla-rust-driver/tree/queryall-backup) branch in case we wanted it in the future
* Removed new `BadQuery` variant, `Connection::query_all` now uses `ProtocolError` to report bad usage.
